### PR TITLE
fix(validation): fix nested portable text markDefs paths

### PR DESCRIPTION
**Pull request by ricokahler**
### Description

The following refactors and fixes how the markDefs in portable text are validated. This work continues off of @skogsmaskin's work in #2931.

Previously `markDefs` were validated inside of the `blockValidator` function. This PR moves that logic into the recursive`validateItem` function. This was done in order to fix two pre-existing bugs:

1. incorrect or repeat paths due to the incorrect path context
2. nested validation for markDefs (e.g. a mark def in a mark def)

This PR also:
1. adds the compiled block schema types to `@sanity/types` + type predicates
2. re-routes `Rule.block` calls to `Rule.custom` and deprecates `Rule.block`

### What to review

Try this out in the portable text editor. Does everything work as expected?

Also, review the tests to see if the assertions are correct.

### Notes for release

Fixes nested validation inside of `markDefs` in the portable text editor.

### CODE REVIEW

File: packages/@sanity/types/src/portableText/types.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
-export interface Block<O = Span> {
+export interface Block<TChild = Span> {
   _type: 'block'
```
Comment (skogsmaskin) [inline packages/@sanity/types/src/portableText/types.ts:2]: This `_type: 'block'` should be string as well as we now supports custom block type names.

**Review by skogsmaskin — COMMENTED** [review]
Great work! I just noticed one thing which I commented on that we should look into.

File: packages/@sanity/types/src/portableText/types.ts (hunk 1 above)
Comment (skogsmaskin) [reply packages/@sanity/types/src/portableText/types.ts:2]: I can confirm that it's not validating properly on hoisted block types.

Comment (skogsmaskin) [reply packages/@sanity/types/src/portableText/types.ts]: I changed it.

**Review by skogsmaskin — APPROVED** [review]
I added some fixes that makes validation work on hoisted (named) block types as well. Please have a look @ricokahler.

**Comment by ricokahler** [comment]
@skogsmaskin ah nice catch! looks good to me. feel free to take over merging etc

**Pull request merged**